doc/man/gnunet-search.1
.TH GNUNET-SEARCH "1" "6 Dec 2009" "GNUnet"
.SH NAME
gnunet\-search \- a command line interface to search for content on GNUnet
.SH SYNOPSIS
.B gnunet\-search
[\fIOPTIONS\fR] [+]\fIKEYWORD\fR [[+]\fIKEYWORD\fR]*

.B gnunet\-search
[\fIOPTIONS\fR] [+]\fIURI\fR
.SH DESCRIPTION
.PP
Search for content on GNUnet. The keywords are case\-sensitive.  gnunet\-search can be used both for a search in the global namespace as well as for searching a private subspace.
.TP
\fB\-a \fILEVEL\fR, \fB\-\-anonymity=\fILEVEL\fR
The \fB-a\fR option can be used to specify additional anonymity constraints.  If set to 0, GNUnet will try to download the file as fast as possible without any additional slowdown for anonymous routing.  Note that you may still have some amount of anonymity depending on the current network load and the power of the adversary.  Use at least 1 to force GNUnet to use anonymous routing.

This option can be used to limit requests further than that. In particular, you can require GNUnet to have a certain amount of cover traffic from other peers before sending your queries. This way, you can gain very high levels of anonymity \- at the expense of much more traffic and much higher latency. So set this option to values beyond 1 only if you really believe you need it.

The definition of ANONYMITY-RECEIVE is the following: If the value v is 0, anonymous routing is not required.  For 1, anonymous routing is required, but there is no lower bound on how much cover traffic must be present.  For values > 1 and < 1000, it means that if GNUnet routes n bytes of messages from foreign peers, it may originate n/v bytes of queries in the same time\-period. The time\-period is twice the average delay that GNUnet deferrs forwarded queries. If the value v is >= 1000, it means that if GNUnet routes n bytes of QUERIES from at least (v % 1000) peers, it may originate n/v/1000 bytes of queries in the same time\-period.

The default is 1 and this should be fine for most users.  Also notice that if you choose values above 1000, you may end up having no throughput at all, especially if many of your fellow GNUnet\-peers do the same.

.TP
\fB\-c \fIFILENAME\fR, \fB\-\-config=\fIFILENAME\fR
use config file (defaults: ~/.gnunet/gnunet.conf)
.TP
\fB\-h\fR, \fB\-\-help\fR
print help page
.TP
\fB\-H \fIHOSTNAME\fR, \fB\-\-host=\fIHOSTNAME\fR
on which host is gnunetd running (default: localhost).  You can also specify a port using the syntax HOSTNAME:PORT.  The default port is 2087.
.TP
\fB\-L \fILOGLEVEL\fR, \fB\-\-loglevel=\fILOGLEVEL\fR
Change the loglevel.  Possible values for LOGLEVEL are NOTHING, FATAL,
ERROR, WARNING, INFO, STATUS and DEBUG.  Note that options in the 
configuration file take precedence over this option (the argument 
will be ignored in that case).
.TP
\fB\-o \fIPREFIX\fR, \fB\-\-output=\fIPREFIX\fR
Writes the encountered (unencrypted) RBlocks or SBlocks to files with name PREFIX.XXX, where XXX is a number. This is useful to keep search results around.
.TP
\fB\-v\fR, \fB\-\-version\fR
print the version number
.SH NOTES
As most GNUnet command\-line tools, gnunet\-search supports passing arguments using environment variables.  This can improve your privacy since otherwise the search terms will likely be visible to other local users.  Setting "GNUNET_ARGS" will cause the respective string to be appended to the actual command\-line and to be processed the same way as arguments given directly at the command line.

You can run gnunet\-search with an URI instead of a keyword.  The URI can have the format for a namespace search or for a keyword search.  For a namespace search, the format is gnunet://ecrs/sks/NAMESPACE/IDENTIFIER.  For a keyword search, use gnunet://ecrs/ksk/KEYWORD[+KEYWORD]*.  If the format does not correspond to a GNUnet URI, GNUnet will automatically assume that keywords are supplied directly.  

If multiple keywords are passed, gnunet\-search will look for content matching any of the keywords.  The prefix "+" makes a keyword mandatory.  

$ gnunet\-search "Das Kapital"

searches for content matching the keyword "Das Kapital".  Whereas

$ gnunet\-search +Das +Kapital

Searches for content matching both mandatory keywords "Das" and "Kapital".

Search results are printed by gnunet\-search like this:
.P
.ad l
	gnunet\-download \-o "COPYING" gnunet://ecrs/chk/HASH1.HASH2.SIZE

		Description: The GNU Public License

		Mime-type: text/plain
.ad b

The first line contains the command to run to download the file.  The suggested filename in the example is COPYING.  The GNUnet URI consists of the key and query hash of the file and finally the size of the file.  After the command to download the file GNUnet will print meta\-data about the file as advertised in the search result, here "The GNU Public License" and the mime\-type (see the options for gnunet\-insert on how to supply meta-data by hand).

.SH FILES
.TP
~/.gnunet/gnunet.conf
GNUnet configuration file; specifies the default value for the timeout
.SH "REPORTING BUGS"
Report bugs by using mantis <https://gnunet.org/bugs/> or by sending electronic mail to <gnunet\-developers@gnu.org>
.SH "SEE ALSO"
\fBgnunet\-gtk\fP(1), \fBgnunet\-insert\fP(1), \fBgnunet\-download\fP(1), \fBgnunet\-pseudonym\fP(1), \fBgnunet.conf\fP(5), \fBgnunetd\fP(1)


ChangeLog
Mon Nov 16 09:25:47 CET 2009
	Allow passing of command-line arguments using GNUNET_ARGS
	environment variable (see Mantis #1467).

Mon Apr 27 22:18:19 CEST 2009
	Monitor low-traffic connections over a longer period
	of time before terminating them.

Fri Feb 27 22:05:50 MST 2009
	Releasing 0.8.0c.

Sun Feb  1 19:51:40 MST 2009
	Fixed Mantis #1429 (struct padding could cause sizeof
	to be different on AMD64, using gcc-ism "packed" to
	avoid this problem).

Sun Feb  1 00:37:16 MST 2009
	Fixed problem in DHT client protocol that could cause
	PUT requests from clients to be (frequently) lost.

Sun Nov  2 16:06:47 MST 2008
	Fixed problem with MySQL disconnects in one module
	impacting MySQL correctness in another module (caused
	crashes).

Sat Aug 16 00:18:34 CEST 2008
	Improved keyword extraction
	Fixed expiration time of keywords
	Releasing 0.8.0b.

Sun Jul 20 15:27:11 MDT 2008
	Releasing 0.8.0a.

Thu Jul 10 00:59:31 MDT 2008
	Fixed bug where datastore resizing would not 
	result in bloomfilter update during gnunet-update.

Sat Jun 28 23:43:00 MDT 2008
	Fixed bug preventing namespace advertisements from
	being processed.

Sat Jun 14 00:54:39 MDT 2008
	Releasing 0.8.0.

Fri Jun 13 23:51:00 MDT 2008
	GNUnet now compiles on OpenBSD.

Wed Jun 11 22:17:09 MDT 2008
	Made state of gnunet-auto-share persistent. 
	Added support for multiple directories.  Improved
	FSUI/ECRS APIs to reduce the number of threads needed
	for probing.  Increased number of concurrent probes
	allowed.

Sat Jun  7 17:54:49 MDT 2008
	Simplified namespaces a lot.  Shorter timeouts for
	trying to connect to gnunetd to prevent interactive
	apps from blocking too long when gnunetd is not
	running.

Sun Jun  1 21:11:54 MDT 2008
	Releasing 0.8.0pre1.

Sun Jun  1 13:41:27 MDT 2008
	Split of libgnunetpseudonym from libgnunetnamespace
	(with major API extension and changes).

Tue May  6 04:10:02 MDT 2008
	Added simple sanity check for [MODULES] values.
	Added "-s" and "-q" options to gnunet-peer-info.

Thu Apr 24 00:05:17 MDT 2008
	Simplified FSUI stop/abort/pause/resume API calls
	(no need to pass FSUI context anymore).

Mon Apr 21 21:05:20 MDT 2008
	Adding buffered IO for FSUI's serialize and
	deserialize code.

Sun Apr 20 20:01:20 MDT 2008
	Fixed bug in HTTP transport causing lonely messages
	(responses to GET) not to be delivered.

Mon Mar 24 21:15:36 MDT 2008
	Fixed various problems with downloading locally
	indexed large files (downloading large files from
	other peers should not have been a problem).
	
Sun Mar  2 15:33:41 MST 2008
	Added option to allow user to disable IPv6 support.
	Releasing 0.8.0pre0.

Sat Mar  1 21:14:02 MST 2008
	Fixed some dstore performance issues.
	Made transports alternate between trying IPv4 and IPv6
	if both are available (since one of them maybe
	misconfigured).

Mon Feb 25 23:25:48 MST 2008
	Cleanup of the DNS code: removed GNUNET_IPvXAddress
	and GNUNET_PRIP; centralized all DNS lookup code in 
	util/network/dns.c; improved IPv6 support for 
        network-client code.

Mon Feb 25 00:01:27 MST 2008
	Added asynchronous search and download methods for 
	ECRS library.  FSUI now can do with only one thread
	per search or download (until now, we had two 
	threads per search / download).

Tue Feb 19 20:35:28 MST 2008
	Updated database schemata to support O(1) operations
	even if there are N files under the same keywords.
	This breaks database backwards compatibility (and
	no migration code is provided).

Mon Feb 18 19:47:37 MST 2008
	Added new options -u and -s to gnunet-insert.

Sat Feb 16 21:37:33 MST 2008
	Implemented gnunet-auto-share for automatic
	sharing of directories.

Thu Feb 14 01:02:32 MST 2008
	Made HTTP transport work over IPv6 as well.

Wed Feb 13 22:59:07 MST 2008
	Modified mySQL and SQLite datastores to return diverse
	result sets during (partial) hash-based get iterations
	(should result in more diverse keyword result sets).

Tue Feb 12 23:54:34 MST 2008
	First shot at integrated hostlist HTTP server based
	on libmicrohttpd.

Sat Feb  9 22:06:23 MST 2008
	Minor changes to the ECRS/FSUI APIs for searches
	(no more timeouts and/or max-results).  Clients who
	need these features could easily implement them 
	directly (and they are not really useful to begin
	with).  
	Also, the IPv4 and IPv6 protocols are being integrated
	so that we only listen to one port (running both IPv4
	and IPv6 over the same port).  This break transport
	compatibility, but given that we're breaking protocol
	compatibility for file-sharing already, this may 
	actually be a bit of a good thing (since it will now
	be possible to tell if peers running the new GAP code
	are connected to other peers running the new GAP code).
	
Sun Feb  3 13:17:09 MST 2008
	Dramatic changes to the GAP implementation (breaking
	protocol compatibility).  Essentially, we can save
	a few bytes in each reply.  More importantly, the new
	code allows the searching client to specify a set of
	replies that are not desired (hopefully helping to
	dramatically increase the diversity of search replies
	obtained over time).  Note that the actual encoding
	and databases are not affected (just P2P protocol).
	The update is not complete yet, but should compile.

Tue Jan  8 20:07:20 MST 2008
	Added option for testing ("make check") to use weak(er)
	PRNG for key generation (thanks to Werner Koch for 
	pointing out how to do it).

Thu Dec 22 20:10:37 MST 2007
	Releasing GNUnet 0.7.3.

Sun Dec  9 14:34:32 MST 2007
	Implemented MySQL version of the dstore-module.  This
	means that sqLite is now again truly optional.

Sat Dec  8 15:15:53 MST 2007
	Integrated F2F topology into main topology module 
	(options are used to choose between F2F-only (old F2F), 
	 minimum number of friend connections (NEW) or entirely
	 arbitrary connection set (default)).

Thu Dec  6 20:51:17 MST 2007
	The SMTP transport is working again.

Thu Nov 22 20:49:16 MST 2007
	United all libgnunetutil_*.so libraries into one big
	libgnunetutil.so library (eliminates issues for binary

src/applications/fs/tools/gnunet-search.c
/*
     This file is part of GNUnet.
     (C) 2001, 2002, 2003, 2004, 2005, 2006 Christian Grothoff (and other contributing authors)

     GNUnet is free software; you can redistribute it and/or modify
     it under the terms of the GNU General Public License as published
     by the Free Software Foundation; either version 2, or (at your
     option) any later version.

     GNUnet is distributed in the hope that it will be useful, but
     WITHOUT ANY WARRANTY; without even the implied warranty of
     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
     General Public License for more details.

     You should have received a copy of the GNU General Public License
     along with GNUnet; see the file COPYING.  If not, write to the
     Free Software Foundation, Inc., 59 Temple Place - Suite 330,
     Boston, MA 02111-1307, USA.
*/

/**
 * @file applications/fs/tools/gnunet-search.c
 * @brief Main function to search for files on GNUnet.
 * @author Christian Grothoff
 */

#include "platform.h"
#include "gnunet_directories.h"
#include "gnunet_fsui_lib.h"
#include "gnunet_util.h"

#define DEBUG 1

static struct GNUNET_GE_Context *ectx;

static struct GNUNET_GC_Configuration *cfg;

static unsigned int anonymity = 1;

static char *cfgFilename = GNUNET_DEFAULT_CLIENT_CONFIG_FILE;

static char *output_filename;

static int errorCode;

static GNUNET_ECRS_FileInfo *fis;

static unsigned int fiCount;


static int
itemPrinter (EXTRACTOR_KeywordType type, const char *data, void *closure)
{
  printf ("\t%20s: %s\n",
          dgettext (LIBEXTRACTOR_GETTEXT_DOMAIN,
                    EXTRACTOR_getKeywordTypeAsString (type)), data);
  return GNUNET_OK;
}

static void
printMeta (const struct GNUNET_MetaData *meta)
{
  GNUNET_meta_data_get_contents (meta, &itemPrinter, NULL);
}

/**
 * Handle the search result.
 */
static void *
eventCallback (void *cls, const GNUNET_FSUI_Event * event)
{
  char *uri;
  char *filename;

#if DEBUG
  GNUNET_EncName enc;
#endif
  switch (event->type)
    {
    case GNUNET_FSUI_search_aborted:
      errorCode = 4;
      GNUNET_shutdown_initiate ();
      break;
    case GNUNET_FSUI_search_result:
      /* retain URIs for possible directory dump later */
      GNUNET_array_grow (fis, fiCount, fiCount + 1);
      fis[fiCount - 1].uri =
        GNUNET_ECRS_uri_duplicate (event->data.SearchResult.fi.uri);
      fis[fiCount - 1].meta =
        GNUNET_meta_data_duplicate (event->data.SearchResult.fi.meta);

      uri = GNUNET_ECRS_uri_to_string (event->data.SearchResult.fi.uri);
#if DEBUG
      if (GNUNET_ECRS_uri_test_loc (event->data.SearchResult.fi.uri))
        {
	  GNUNET_PeerIdentity id;
          GNUNET_ECRS_uri_get_peer_identity_from_loc (event->data.SearchResult.fi.uri,
						      &id);
	  GNUNET_hash_to_enc (&id.hashPubKey, &enc);
          fprintf (stdout,
		   "Received locURI putting data at peer %s\n",
                   (const char*) &enc);
        }
#endif
      printf ("%s:\n", uri);
      filename =
        GNUNET_meta_data_get_by_type (event->data.SearchResult.fi.meta,
                                      EXTRACTOR_FILENAME);
      if (filename != NULL)
        {
          char *dotdot;

          while (NULL != (dotdot = strstr (filename, "..")))
            dotdot[0] = dotdot[1] = '_';

          printf ("gnunet-download -o \"%s\" %s\n", filename, uri);
        }
      else
        printf ("gnunet-download %s\n", uri);
      printMeta (event->data.SearchResult.fi.meta);
      printf ("\n");
      fflush (stdout);
      GNUNET_free_non_null (filename);
      GNUNET_free (uri);
      break;
    case GNUNET_FSUI_search_started:
    case GNUNET_FSUI_search_stopped:
    case GNUNET_FSUI_search_update:
      break;
    default:
      GNUNET_GE_BREAK (NULL, 0);
      break;
    }
  return NULL;
}

/**
 * All gnunet-search command line options
 */
static struct GNUNET_CommandLineOption gnunetsearchOptions[] = {
  {'a', "anonymity", "LEVEL",
   gettext_noop ("set the desired LEVEL of sender-anonymity"),
   1, &GNUNET_getopt_configure_set_uint, &anonymity},
  GNUNET_COMMAND_LINE_OPTION_CFG_FILE (&cfgFilename),   /* -c */
  GNUNET_COMMAND_LINE_OPTION_HELP (gettext_noop ("Search GNUnet for files.")),  /* -h */
  GNUNET_COMMAND_LINE_OPTION_HOSTNAME,  /* -H */
  GNUNET_COMMAND_LINE_OPTION_LOGGING,   /* -L */
  {'o', "output", "FILENAME",
   gettext_noop ("write encountered (decrypted) search results to FILENAME"),
   1, &GNUNET_getopt_configure_set_string, &output_filename},
  GNUNET_COMMAND_LINE_OPTION_VERSION (PACKAGE_VERSION), /* -v */
  GNUNET_COMMAND_LINE_OPTION_VERBOSE,
  GNUNET_COMMAND_LINE_OPTION_END,
};

/**
 * The main function to search for files on GNUnet.
 *
 * @param argc number of arguments from the command line
 * @param argv command line arguments
 * @return return value from gnunet-search: 0: ok, -1: error
 */
int
main (int argc, char *const *argv)
{
  struct GNUNET_ECRS_URI *uri;
  int i;
  struct GNUNET_FSUI_Context *ctx;
  struct GNUNET_FSUI_SearchList *s;

  i = GNUNET_init (argc,
                   argv,
                   "gnunet-search [OPTIONS] [KEYWORDS]",
                   &cfgFilename, gnunetsearchOptions, &ectx, &cfg);
  if (i == GNUNET_SYSERR)
    {
      GNUNET_fini (ectx, cfg);
      return -1;
    }
  /* convert args to URI */
  uri =
    GNUNET_ECRS_keyword_command_line_to_uri (ectx, argc - i,
                                             (const char **) &argv[i]);
  if (uri == NULL)
    {
      printf (_("Error converting arguments to URI!\n"));
      errorCode = -1;
      goto quit;
    }
  ctx =
    GNUNET_FSUI_start (ectx, cfg, "gnunet-search", 4, GNUNET_NO,
                       &eventCallback, NULL);
  if (ctx == NULL)
    {
      GNUNET_ECRS_uri_destroy (uri);
      GNUNET_fini (ectx, cfg);
      return GNUNET_SYSERR;
    }
  errorCode = 1;
  s = GNUNET_FSUI_search_start (ctx, anonymity, uri);
  GNUNET_ECRS_uri_destroy (uri);
  if (s == NULL)
    {
      errorCode = 2;
      GNUNET_FSUI_stop (ctx);
      goto quit;
    }
  GNUNET_shutdown_wait_for ();
  if (errorCode == 1)
    GNUNET_FSUI_search_abort (s);
  GNUNET_FSUI_search_stop (s);
  GNUNET_FSUI_stop (ctx);

  if (output_filename != NULL)
    {
      char *outfile;
      unsigned long long n;
      char *data;
      struct GNUNET_MetaData *meta;

      meta = GNUNET_meta_data_create ();
      /* ?: anything here to put into meta? */
      if (GNUNET_OK ==
          GNUNET_ECRS_directory_create (ectx, &data, &n, fiCount, fis, meta))
        {
          outfile = GNUNET_expand_file_name (ectx, output_filename);
          GNUNET_disk_file_write (ectx, outfile, data, n, "600");
          GNUNET_free (outfile);
          GNUNET_free (data);
        }
      GNUNET_free (output_filename);
    }
  for (i = 0; i < fiCount; i++)
    {
      GNUNET_ECRS_uri_destroy (fis[i].uri);
      GNUNET_meta_data_destroy (fis[i].meta);
    }
  GNUNET_array_grow (fis, fiCount, 0);
quit:
  GNUNET_fini (ectx, cfg);
  return errorCode;
}

/* end of gnunet-search.c */
